Fix permission error on Windows when ESTAT returns footer with URL to zipfile

Running the test suite on Windows (no remote-data suffices) raises a permission error:

```
tests\test_sources.py:168:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pandasdmx\api.py:420: in get
    msg = self.source.finish_message(msg, self, **kwargs)
pandasdmx\source\estat.py:66: in finish_message
    return request.get(url=url, tofile=ntf.name)
pandasdmx\api.py:394: in get
    response_content = remote.ResponseIO(response, *arg)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <pandasdmx.remote.ResponseIO object at 0x000002743AED7EC8>, response = <Response [200]>
tee = 'C:\\Users\\a3305\\AppData\\Local\\Temp\\pandasdmx-zbrnkoip'

    def __init__(self, response, tee=None):
        self.response = response
        self.chunks = response.iter_content(ITER_CHUNK_SIZE)
        self.pending = bytes()
        self.tee_filename = tee
        # str() here is for Python 3.5 compatibility
>       self.tee = open(str(tee), 'wb') if tee else None
E       PermissionError: [Errno 13] Permission denied: 'C:\\Users\\a3305\\AppData\\Local\\Temp\\pandasdmx-zbrnkoip'

pandasdmx\remote.py:79: PermissionError
_______________________
```
For more background see [here](https://stackoverflow.com/questions/23212435/permission-denied-to-write-to-my-temporary-file)

Solution: Make remote.ResponseIO accept a file-like object for the tee arg and have source.estat.finish_message (line 66) pass the tempfile rather than its name. Adjust api.Request.get accordingly.

--- a/pandasdmx/api.py
+++ b/pandasdmx/api.py
@@ -388,8 +388,7 @@ class Request:
                 raise
 
         # Maybe copy the response to file as it's received
-        arg = [tofile] if tofile else [] # explain or remove this. Why wrap in list and unpack it in next line?
-        response_content = remote.ResponseIO(response, *arg)
+        response_content = remote.ResponseIO(response, tee=tofile)
 
         # Select reader class
         content_type = response.headers.get('content-type', None)

--- a/pandasdmx/api.py
+++ b/pandasdmx/api.py
@@ -277,8 +277,8 @@ class Request:
             Type of resource to retrieve.
         resource_id : str, optional
             ID of the resource to retrieve.
-        tofile : str or :class:`~os.PathLike`, optional
-            File path to write SDMX data as it is recieved.
+        tofile : str or :class:`~os.PathLike` or `file-like object`, optional
+            File path or file-like to write SDMX data as it is recieved.
         use_cache : bool, optional
             If :obj:`True`, return a previously retrieved :class:`~.Message`
             from :attr:`cache`, or update the cache with a newly-retrieved
@@ -388,7 +388,7 @@ class Request:
                 raise
 
         # Maybe copy the response to file as it's received
-        arg = [tofile] if tofile else []
+        arg = [tofile] if tofile else [] # explain or remove this. Why wrap in list and unpack it in next line?
         response_content = remote.ResponseIO(response, *arg)
 
         # Select reader class

--- a/pandasdmx/remote.py
+++ b/pandasdmx/remote.py
@@ -92,7 +92,7 @@ class ResponseIO(BufferedIOBase):
 
         If the argument is omitted, :py:obj:`None`, or negative, reads and
         returns all data until EOF. If *tee* was provided to the constructor,
-        data is echoed to file. file is not closed.
+        data is echoed to file; *tee* is not closed.
 
         If the argument is positive, and the underlying raw stream is not
         ‘interactive’, multiple raw reads may be issued to satisfy the byte

--- a/pandasdmx/remote.py
+++ b/pandasdmx/remote.py
@@ -64,7 +64,7 @@ class ResponseIO(BufferedIOBase):
     ----------
     response : :class:`requests.Response`
         HTTP response to wrap.
-    tee : :py:class:`os.PathLike`, optional
+    tee : :py:class:`os.PathLike` or :py:class:`io.BufferedIOBasè`, optional
         If provided, the contents of the response are also written to this file
         as they are received. The file is closed automatically when *response*
         reaches EOF.
@@ -75,9 +75,14 @@ class ResponseIO(BufferedIOBase):
         self.response = response
         self.chunks = response.iter_content(ITER_CHUNK_SIZE)
         self.pending = bytes()
-        self.tee_filename = tee
-        # str() here is for Python 3.5 compatibility
-        self.tee = open(str(tee), 'wb') if tee else None
+        # check for file-like or tempfile
+        if isinstance(tee, BufferedIOBase) or hasattr(tee, 'file'):
+            self.tee_filename = tee.name
+            self.tee = tee
+        else:
+            # so tee must be str or os.PathLike or None
+            self.tee_filename = tee
+            self.tee = open(tee, 'wb') if tee else None
 
     def readable(self):
         return True
@@ -87,7 +92,7 @@ class ResponseIO(BufferedIOBase):
 
         If the argument is omitted, :py:obj:`None`, or negative, reads and
         returns all data until EOF. If *tee* was provided to the constructor,
-        data is echoed to file.
+        data is echoed to file. file is not closed.
 
         If the argument is positive, and the underlying raw stream is not
         ‘interactive’, multiple raw reads may be issued to satisfy the byte
@@ -116,8 +121,5 @@ class ResponseIO(BufferedIOBase):
 
         if self.tee:
             self.tee.write(result)
-            if len(result) == 0:
-                # Also close the results file
-                self.tee.close()
 
         return result

--- a/pandasdmx/source/estat.py
+++ b/pandasdmx/source/estat.py
@@ -57,18 +57,18 @@ class Source(BaseSource):
         wait_seconds, attempts = get_footer_url
 
         # Create a temporary file to store the ZIP response
-        with NamedTemporaryFile(prefix='pandasdmx-') as ntf:
-            # Make a limited number of attempts to retrieve the file
-            for a in range(attempts):
-                sleep(wait_seconds)
-                try:
-                    # This line succeeds if the file exists; the ZIP response
-                    # is stored to ntf.name, and then used by the
-                    # handle_response() hook below
-                    return request.get(url=url, tofile=ntf.name)
-                except requests.HTTPError:
-                    raise
-
+        ntf = NamedTemporaryFile(prefix='pandasdmx-')
+        # Make a limited number of attempts to retrieve the file
+        for a in range(attempts):
+            sleep(wait_seconds)
+            try:
+                # This line succeeds if the file exists; the ZIP response
+                # is stored to ntf, and then used by the
+                # handle_response() hook below
+                return request.get(url=url, tofile=ntf)
+            except requests.HTTPError:
+                raise
+        ntf.close()
         raise RuntimeError('Maximum attempts exceeded')
 
     def handle_response(self, response, content):
@@ -79,6 +79,7 @@ class Source(BaseSource):
         contained XML file.
 
         """
+        
         if response.headers['content-type'] != 'application/octet-stream':
             return response, content
 
@@ -89,7 +90,7 @@ class Source(BaseSource):
                 break
 
         # Open the zip archive
-        with ZipFile(content.tee_filename, mode='r') as zf:
+        with ZipFile(content.tee, mode='r') as zf:
             # The archive should contain only one file
             infolist = zf.infolist()
             assert len(infolist) == 1

--- a/tests/test_sources.py
+++ b/tests/test_sources.py
@@ -260,7 +260,8 @@ class TestINSEE(DataSourceTest):
         req.get(endpoint, provider='all',
                 tofile=self._cache_path.with_suffix('.' + endpoint))
 
-
+@pytest.mark.xfail(reason='Service is currently unavailable.',
+                   raises=HTTPError)
 class TestISTAT(DataSourceTest):
     source_id = 'ISTAT'
 

